File: 1506-find-root-of-n-ary-tree/1506-find-root-of-n-ary-tree.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    
    public Node() {
        children = new ArrayList<Node>();
    }
    
    public Node(int _val) {
        val = _val;
        children = new ArrayList<Node>();
    }
    
    public Node(int _val,ArrayList<Node> _child0ren) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    public Node findRoot(List<Node> tree) {
        HashSet<Integer> childrenSet = new HashSet<>();
        //add every nodes all the children in our set.
        for(Node node : tree) {
            for(Node child : node.children) {
                childrenSet.add(child.val);
            }
        }
        Node root = null;
        //check which node is not present in children set, that node is our root node.
        for(Node node : tree) {
            if(!childrenSet.contains(node.val)) root = node;
        }
        
        return root;
    }
}

File: 1506-find-root-of-n-ary-tree/1506-find-root-of-n-ary-tree.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
public:
    Node* findRoot(vector<Node*> tree) {
        unordered_set<int> childrenSet;
        //loop over trees to add all the children of each node.
        for(auto* node : tree) {
            for(auto* child : node->children) {
                childrenSet.insert(child->val);
            }
        }
        //check which node in tree vector is not present in our childrenSet
        Node* root = nullptr;
        for(auto* node : tree) {
            if(childrenSet.find(node->val) == childrenSet.end()) root = node;
        }
        
        return root;
    }
};